### DISCUSSION TITLE
ulauncher is closes after switching keyboard layout

### DISCUSSION THREAD

**Discussion by uncletoxa** [Ulauncher/Ulauncher#891] [General]
### Steps to reproduce
1. ctrl+space to open ulauncher window
2. alt+shift to change keyboard layout to type in different language

### Current behavior
ulauncher window closes

### Expected behavior
ulauncher window is still open waiting for input

### Environment
**Ulauncher version**: 4.1.0.r1
**Linux distro**: Ubuntu 18.04 on X11
**Desktop environment**: gnome 

### Logs
```
/usr/lib/python2.7/dist-packages/ulauncher/util/desktop/notification.py:1: PyGIWarning: Notify was imported without specifying a version first. Use gi.require_version('Notify', '0.7') before import to ensure that the right version gets loaded.
  from gi.repository import Notify
/usr/lib/python2.7/dist-packages/ulauncher/ui/windows/PreferencesUlauncherDialog.py:6: PyGIWarning: WebKit2 was imported without specifying a version first. Use gi.require_version('WebKit2', '4.0') before import to ensure that the right version gets loaded.
  from gi.repository import Gio, Gtk, WebKit2, GLib
```

**Comment by friday** [comment]
Keyboard shortcuts that draws an OSD, like [Gnome does for this scenario](https://blogs.gnome.org/mclasen/2012/12/18/input-sources-in-gnome-3-7-4/) causes that OSD to steal the focus from Ulauncher, which then closes. This is the intended behavior when losing focus. I see why you wouldn't want it to in this instance, but I doubt it's even possible to make exceptions for specific OSDs.

You could create a custom switcher without the OSD and map it to your current key combo: https://askubuntu.com/q/209597/420716, or if you always want to use English for Ulauncher for example, you could create a custom wrapper for `ulauncher-toggle` so that it switches before opening, then switches back after.

**Reply by paulelms** [reply]
When switching via xorg (without gnome osd), the launcher closes anyway

**Comment by gornostal** [comment]
Thanks for the detailed answer @friday 

@ToxaZ sorry, I'm going to have to close this one as there's no feasible fix for it.

**Comment by uncletoxa** [comment]
Could you please tell me the way to add wrapper to ulauncher-toggle? Should I clone the repo and rebuild the project?

**Comment by friday** [comment]
Do you know how to write a shell script?
You can wrap any executable with another executable. `ulauncher-toggle` is already a wrapper, so you can copy it to another location in `$PATH` or rename it, then modify it how you want.

You don't want to modify the original file since it will be overwritten by your package manager when Ulauncher updates.

**Comment by uncletoxa** [comment]
If I'd need to switch layout before launching ulauncher, that'll be easy like
```
CURRENT_KEYMAP=$(setxkbmap -print | awk -F"+" '/xkb_symbols/ {print $2}') 

setxkbmap us
/usr/lib/ulauncher
setxkbmap "$CURRENT_KEYMAP"
```

But I don't understand how to wrap a call to already running program.

**Comment by friday** [comment]
> But I don't understand how to wrap a call to already running program.

You're not supposed to wrap Ulauncher, but `ulauncher-toggle`. You can wrap any command, including compiled binaries and running programs, as long as you change the way they open to point to your wrapper or make the wrapper override them using your `$PATH` order.

`ulauncher-toggle` is a shells script, so you can open it in a text editor and add `setxkbmap us` after `#!/bin/sh`. Then save as a new binary and make it executable. If this doesn't work, but calling the wrapper from the terminal does, then make sure the keyboard combination is mapped to trigger the new binary location (I'm not sure but [you may need to do this](https://github.com/Ulauncher/Ulauncher/wiki/Hotkey-In-Wayland)).

Restoring the old keyboard layout after closing Ulauncher is harder. You may have to write a loop and check that Ulauncher isn't the active window, then restore.

For example `xdotool getactivewindow getwindowname` should print `Ulauncher window title` while Ulauncher is open. There are probably other ways without xdotool.

**Comment by uncletoxa** [comment]
**UPD:** Won't work on ver. >5

For smb encountering the same issue. 
Best solution for me was to use gnome DBus api (using `setxkbmap` gives a trouble switching back to another layout) calling a ulauncher-toggle:
```
#!/bin/sh

gdbus call \
    --session --dest org.gnome.Shell \
    --object-path /org/gnome/Shell \
    --method org.gnome.Shell.Eval \
    "imports.ui.status.keyboard.getInputSourceManager().inputSources[0].activate()"
    # [0] on prev line is the index of layout you want to switch, change it to [1] for second layout, [2] for third, etc

dbus-send \
    --session \
    --print-reply \
    --dest=net.launchpad.ulauncher \
    /net/launchpad/ulauncher \
    net.launchpad.ulauncher.toggle_window

# dbus-send is asynchronous and sometimes it takes a bit of time for Ulauncher window to be presented
sleep 0.03

wmctrl -a "Ulauncher window title"

**Comment by trin4ik** [comment]
same here, but i need swith differents layout in ulauncher.
my fix, in file 
`/ulauncher/ui/windows/UlauncherWindow.py`
change
```python
    def on_focus_out_event(self, widget, event):
        # apparently Gtk doesn't provide a mechanism to tell if window is in focus
        # this is a simple workaround to avoid hiding window
        # when user hits Alt+key combination or changes input source, etc.
        self.is_focused = False
        t = threading.Timer(0.07, lambda: self.is_focused or self.hide())
        t.start()

```
to
```python
    def on_focus_out_event(self, widget, event):
        # apparently Gtk doesn't provide a mechanism to tell if window is in focus
        # this is a simple workaround to avoid hiding window
        # when user hits Alt+key combination or changes input source, etc.
        self.is_focused = False
        t = threading.Timer(.2, lambda: self.is_focused or self.hide())
        t.start()

```
and i can switch layout without blur event on ulauncher.

**Comment by 0Slmpnv0** [comment]
Has anyone solved this problem? Or maybe there are some launchers, that don't have it?

**Reply by ZhengshuaiPENG** [reply]
Same in Linux Mint， I tried Albert before，it works，I am searching the solution now

**Comment by friday** [comment]
I don't think any launcher can solve this without making assumptions about the environment that won't hold everywhere (especially Wayland), using hacks or disabling the focus out handler completely.

1. If we try to detect if the new window is an OSD rather than a window, we could *maybe* do that on X11, but not Wayland.
1. If we add back the workaround to wait and see if the focus is restored within a fixed duration, then we reintroduce a race condition that only works for some, and potentially make the focus out behavior feel slow for everyone. The old 70ms timeout or even 0.2s won't be enough for the Gnome keybord layout switching OSD. If you have more than two languages you need a lot longer.
1. If we try to detect and change the layout programmatically we will only be able to support this for [specific Desktops](https://www.linuxquestions.org/questions/linux-desktop-74/setting-keyboard-layouts-programmatically-4175686841/) via setxkbmap, Gsettings or `ibus` ( see `ibus help` for all `ibus` commands). 

Possible remaining options:
1. Add an option to disable hiding the window completely on focus out.
2. Adding command hooks for opening and closing the window that lets users handle option 3 for their desktop specifically (the second hook could get the output of the first hook as an argument).
3. If we listen to the key press events and try to read the key codes of the keys the user presses and "normalize" it to latin character set. This would be similar to when you press a key at https://www.toptal.com/developers/keycode and get the `event.code` instead of `event.key`. This feels like a bad solution to me though. It would likely only work for some and create new problems.

**Comment by friday** [comment]
I just added option 1 as an experimental setting in ec58eadc

There is no UI for it (because it's experimental). If you run from the v6 branch (will be in the next release after v6.0.0-beta11) and edit `~/.config/ulauncher/settings.json` you can enable it by setting `close_on_focus_out` to `false`.
